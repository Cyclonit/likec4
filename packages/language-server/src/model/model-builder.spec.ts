import { describe, expect, it } from 'vitest'
import { createTestServices } from '../test'
import { keys } from 'rambdax'
import type { Fqn, ViewID } from '@likec4/core/types'

describe('LikeC4ModelBuilder', () => {
  it('builds model', async () => {
    const { validate, buildModel } = createTestServices()
    const { diagnostics } = await validate(`
    specification {
      element component
      element user {
        style {
          shape: person
        }
      }
      tag deprecated
    }
    model {
      user client {
        -> frontend
      }
      component system {
        backend = component 'Backend' {
          technology 'NodeJS'

          style {
            color secondary
          }
        }
        component frontend {
          #deprecated
          description 'Frontend description'

          style {
            color: muted
            shape: browser
          }

          -> backend 'requests'
        }
      }
    }
    `)
    expect(diagnostics).toHaveLength(0)
    const model = await buildModel()
    expect(model).toBeDefined()
    expect(model.elements).toMatchObject({
      client: {
        kind: 'user',
        shape: 'person'
      },
      'system.backend': {
        color: 'secondary',
        title: 'Backend',
        technology: 'NodeJS'
      },
      'system.frontend': {
        color: 'muted',
        shape: 'browser',
        description: 'Frontend description'
      }
    })
    expect(model.elements['client' as Fqn]).not.toHaveProperty('color')
    expect(model.elements['system' as Fqn]).not.toHaveProperty('color')
    expect(model.elements['system' as Fqn]).not.toHaveProperty('shape')
    expect(model.elements['system.backend' as Fqn]).toHaveProperty('color', 'secondary')
    expect(model.elements['system.backend' as Fqn]).not.toHaveProperty('description')

    expect(model).toMatchSnapshot()
  })

  it('builds model with extend', async () => {
    const { parse, validateAll, buildModel } = createTestServices()
    await parse(`
    specification {
      element component
      element user
      tag deprecated
    }
    model {
      user client
      component system {
        backend = component
        component frontend
      }
    }
    `)
    await parse(`
    model {
      extend system.backend {
        component api
      }
      system.frontend -> api 'requests'
      client -> system.frontend {
        title 'opens'
      }
    }
    views {
      view index {
        title 'Index'
        include *
      }

      view v1 of api {
        include *
        autoLayout LR
      }

      view of system.frontend {
        include *
      }
    }
    `)
    const { errors } = await validateAll()
    expect(errors).toEqual([])
    const model = await buildModel()
    expect(model).toBeDefined()
    expect(model.elements).toMatchObject({
      client: {
        kind: 'user'
      },
      'system.backend.api': {
        kind: 'component'
      }
    })
    expect(keys(model.relations)).toHaveLength(2)

    expect(keys(model.views)).toHaveLength(3)
    expect(model.views).toMatchObject({
      index: {
        id: 'index',
        title: 'Index',
        autoLayout: 'TB'
      },
      v1: {
        id: 'v1',
        viewOf: 'system.backend.api',
        title: 'api',
        autoLayout: 'LR'
      }
    })
    expect(model.views['index' as ViewID]).not.toHaveProperty('viewOf')

    expect(model).toMatchSnapshot()
  })
})
