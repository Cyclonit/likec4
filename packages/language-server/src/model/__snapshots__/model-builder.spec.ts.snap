// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LikeC4ModelBuilder > builds model 1`] = `
{
  "elements": {
    "client": {
      "id": "client",
      "kind": "user",
      "shape": "person",
      "title": "client",
    },
    "system": {
      "id": "system",
      "kind": "component",
      "title": "system",
    },
    "system.backend": {
      "color": "secondary",
      "id": "system.backend",
      "kind": "component",
      "technology": "NodeJS",
      "title": "Backend",
    },
    "system.frontend": {
      "color": "muted",
      "description": "Frontend description",
      "id": "system.frontend",
      "kind": "component",
      "shape": "browser",
      "tags": [
        "deprecated",
      ],
      "title": "frontend",
    },
  },
  "relations": {
    "171cae8ab349b2ef3f3d40f23c7a7fa2d46a2b23": {
      "id": "171cae8ab349b2ef3f3d40f23c7a7fa2d46a2b23",
      "source": "system.frontend",
      "target": "system.backend",
      "title": "requests",
    },
    "6f3533077304f2e812d6c9cfdb6be977e23efe3d": {
      "id": "6f3533077304f2e812d6c9cfdb6be977e23efe3d",
      "source": "client",
      "target": "system.frontend",
      "title": "",
    },
  },
  "views": {},
}
`;

exports[`LikeC4ModelBuilder > builds model with extend 1`] = `
{
  "elements": {
    "client": {
      "id": "client",
      "kind": "user",
      "title": "client",
    },
    "system": {
      "id": "system",
      "kind": "component",
      "title": "system",
    },
    "system.backend": {
      "id": "system.backend",
      "kind": "component",
      "title": "backend",
    },
    "system.backend.api": {
      "id": "system.backend.api",
      "kind": "component",
      "title": "api",
    },
    "system.frontend": {
      "id": "system.frontend",
      "kind": "component",
      "title": "frontend",
    },
  },
  "relations": {
    "671e3db5eb605d5f245f247d95925bd0953e0e4c": {
      "id": "671e3db5eb605d5f245f247d95925bd0953e0e4c",
      "source": "system.frontend",
      "target": "system.backend.api",
      "title": "requests",
    },
    "99bffdfd5c78e024393e6223cd9b19225321a5e3": {
      "id": "99bffdfd5c78e024393e6223cd9b19225321a5e3",
      "source": "client",
      "target": "system.frontend",
      "title": "opens",
    },
  },
  "views": {
    "74f5364fa4b064ce9830043b93d4cc6fce9d85b0": {
      "edges": [
        {
          "id": "system.frontend:system.backend",
          "label": "requests",
          "parent": null,
          "relations": [
            "671e3db5eb605d5f245f247d95925bd0953e0e4c",
          ],
          "source": "system.frontend",
          "target": "system.backend",
        },
        {
          "id": "client:system.frontend",
          "label": "opens",
          "parent": null,
          "relations": [
            "99bffdfd5c78e024393e6223cd9b19225321a5e3",
          ],
          "source": "client",
          "target": "system.frontend",
        },
      ],
      "id": "74f5364fa4b064ce9830043b93d4cc6fce9d85b0",
      "nodes": [
        {
          "children": [],
          "color": "primary",
          "id": "client",
          "parent": null,
          "shape": "rectangle",
          "title": "client",
        },
        {
          "children": [],
          "color": "primary",
          "id": "system.backend",
          "parent": null,
          "shape": "rectangle",
          "title": "backend",
        },
        {
          "children": [],
          "color": "primary",
          "id": "system.frontend",
          "parent": null,
          "shape": "rectangle",
          "title": "frontend",
        },
      ],
      "rules": [
        {
          "exprs": [
            {
              "wildcard": true,
            },
          ],
          "isInclude": true,
        },
      ],
      "title": "frontend",
      "viewOf": "system.frontend",
    },
    "index": {
      "edges": [
        {
          "id": "client:system",
          "label": "opens",
          "parent": null,
          "relations": [
            "99bffdfd5c78e024393e6223cd9b19225321a5e3",
          ],
          "source": "client",
          "target": "system",
        },
      ],
      "id": "index",
      "nodes": [
        {
          "children": [],
          "color": "primary",
          "id": "client",
          "parent": null,
          "shape": "rectangle",
          "title": "client",
        },
        {
          "children": [],
          "color": "primary",
          "id": "system",
          "parent": null,
          "shape": "rectangle",
          "title": "system",
        },
      ],
      "rules": [
        {
          "exprs": [
            {
              "wildcard": true,
            },
          ],
          "isInclude": true,
        },
      ],
      "title": "Index",
    },
    "v1": {
      "edges": [
        {
          "id": "system.frontend:system.backend.api",
          "label": "requests",
          "parent": null,
          "relations": [
            "671e3db5eb605d5f245f247d95925bd0953e0e4c",
          ],
          "source": "system.frontend",
          "target": "system.backend.api",
        },
      ],
      "id": "v1",
      "nodes": [
        {
          "children": [],
          "color": "primary",
          "id": "system.frontend",
          "navigateTo": "74f5364fa4b064ce9830043b93d4cc6fce9d85b0",
          "parent": null,
          "shape": "rectangle",
          "title": "frontend",
        },
        {
          "children": [],
          "color": "primary",
          "id": "system.backend.api",
          "parent": null,
          "shape": "rectangle",
          "title": "api",
        },
      ],
      "rules": [
        {
          "exprs": [
            {
              "wildcard": true,
            },
          ],
          "isInclude": true,
        },
      ],
      "title": "api",
      "viewOf": "system.backend.api",
    },
  },
}
`;
